src/pgen/homologous.cpp

```diff
@@ -0,0 +1,137 @@
+      "Problem \"homolohous\" requires \"MonopoleSph\" \" geometry!");
+
+  auto &rc = pmb->meshblock_data.Get();
+
+  PackIndexMap imap;
+  auto v = rc->PackVariables(
+      {fluid_prim::density, fluid_prim::velocity, fluid_prim::energy, fluid_prim::bfield,
+       fluid_prim::ye, fluid_prim::pressure, fluid_prim::temperature, fluid_prim::gamma1},
+      imap);
+
+  const int irho = imap[fluid_prim::density].first;
+  const int ivlo = imap[fluid_prim::velocity].first;
+  const int ivhi = imap[fluid_prim::velocity].second;
+  const int ieng = imap[fluid_prim::energy].first;
+  const int ib_lo = imap[fluid_prim::bfield].first;
+  const int ib_hi = imap[fluid_prim::bfield].second;
+  const int iye = imap[fluid_prim::ye].second;
+  const int iprs = imap[fluid_prim::pressure].first;
+  const int itmp = imap[fluid_prim::temperature].first;
+  const int igm1 = imap[fluid_prim::gamma1].first;
+
+
+  
+  const Real Cv = pin->GetOrAddReal("homologous", "Cv", 1.0);
+  const Real gamma = pin->GetOrAddReal("homologous", "gamma", 1.3334);
+  const bool spherical = pin->GetOrAddReal("homologous", "spherical_coords", true);
+  const string table = pin ->GetOrAddReal("homologous","table");
+  auto &coords = pmb->coords;
+  auto pmesh = pmb->pmy_mesh;
+  int ndim = pmesh->ndim;
+
+  const Real v_inner = (4. / 3.) * M_PI * std::pow(rinner, 3.);
+  const Real uinner = Eexp / v_inner;
+
+  IndexRange ib = pmb->cellbounds.GetBoundsI(IndexDomain::entire);
+  IndexRange jb = pmb->cellbounds.GetBoundsJ(IndexDomain::entire);
+  IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::entire);
+
+
+
+  auto eos = pmb->packages.Get("eos")->Param<singularity::EOS>("d.EOS"); 
+  auto emin = pmb->packages.Get("eos")->Param<Real>("sie_min");
+  auto emax = pmb->packages.Get("eos")->Param<Real>("sie_max");
+
+
+  //Initial Data
+  std::vector<double> mydata[11];
+  readtable(table,mydata);
+
+  //MonopoleGR
+  auto monopole_pkg = pmb->packages.Get("monopole_gr");
+  auto matter = monopolepkg->Param<MonopoleGR::Matter_host_t("matter");
+  auto matter_h = monopolepkg->Param<MonopoleGR::Matter_host_t("matter_h");
+  for (int i = 0; i < npoints; ++i){
+    matter_h(MonopoleGR::Matter::RHO,i)=mydata[COLUMNS::rho_adm][i];
+    matter_h(MonopoleGR::Matter::J_R,i)=mydata[COLUMNS::P_adm][i];
+    matter_h(MonopoleGR::Matter::trcS,i)=mydata[COLUMNS::S_adm][i];
+    matter_h(MonopoleGR::Matter::Srr,i)=mydata[COLUMNS::Srr_adm][i];
+  }
+  Kokkos::deep_copy(matter,matter_h);
+  MonopoleGR::IntegrateHypersurface(monopole_pkg);
+  MonopoleGR::LinearSolveForAlpha(monopole_pkg);
+  MonopoleGR::SpacetimeToDevice(monopole_pkg);
+  MonopoleGR::DumpToTxt("phoebus-metric-setup.dat",monopole_pkg);
+
+  // Primitive Variables
+  auto MoveVectorToDevice(std::vector<double> &vec){
+    Kokkos::View<Real*, Kokkos::HostSpace, Kokkos::MemoryUnmanaged> hm (vec.data(),vec.size());
+    return;
+    Kokkos::create_mirror_view_and_copy(Kokkos::DefaultExecutionSpace::memory_space(),hm);
+  }
```

Comment:
You can't have nested function definitions in C++ in the same way you can in python. This needs to be a fucntion above the `pgen` declaration. OR you can define it inline as a lambda:
```C++
  auto MoveVectorToDevice = [](std::vector<double> &vec) {
    Kokkos::View<Real*, Kokkos::HostSpace, Kokkos::MemoryUnmanaged> hm (vec.data(),vec.size());
    return;
    Kokkos::create_mirror_view_and_copy(Kokkos::DefaultExecutionSpace::memory_space(),hm);
  }
```

Reply:
defined inline as lambda